code/datums/music/music.dm

```diff
@@ -0,0 +1,263 @@
+		var/atom/movable/M = player.parent
+		if(M.z != target_mob.z)
+			continue
+		var/dx = M.x-target_mob.x
+		var/dy = M.y-target_mob.y
+		var/range_squared = dx*dx+dy*dy
+		if(range_squared < closest_range_squared)
+			closest_range_squared = range_squared
+	var/closest_range = sqrt(closest_range_squared)
+	var/target
+	if(closest_range <= full_range)
+		target = 1
+	else if(closest_range >= soft_range)
+		target = 0
+	else
+		target = (soft_range - closest_range) / (soft_range - full_range)
+	if(!is_fading || fade_volume_target != target)
+		fade_at_rate(target, range_fade_speed)
+
+	. = ..()
+
+/datum/music/sourced/Destroy()
+	. = ..()
+	for(var/_player in players)
+		var/datum/component/music_player/player = _player
+		player.mob_players -= target_mob
+
+/datum/component/music_player
+	var/list/mob_players = list()
+	var/music_path
+	/// Whether new mobs will be added. To force mobs to stop playing this, use stop_all()
+	var/enabled = TRUE
+	/// Range within which music will start to play
+	var/start_range = 7
+	var/fade_in_time = 20
+	/// Whether music of the same typepath is shared
+	var/shared = TRUE
+
+/datum/component/music_player/Initialize(_music_path)
+	if(_music_path)
+		music_path = _music_path
+	START_PROCESSING(SSprocessing, src)
+
+/datum/component/music_player/proc/do_range_check(var/fade_time = fade_in_time)
+	if(!music_path)
+		return
+	var/shared = FALSE
+	for(var/mob/living/M in range(start_range, parent))
+		if(!M.client)
+			continue
+		if(mob_players[M])
+			continue
+		var/did_find = FALSE
+		if(shared)
+			for(var/_music in M.client.active_music)
+				var/datum/music/sourced/music = _music
+				if(istype(music, music_path))
+					mob_players[M] = music
+					music.players += src
+					did_find = TRUE
+					break
+		if(!did_find)
+			var/datum/music/sourced/music = new music_path(M.client, fade_time > 0 ? 0 : 1)
+			if(!music.gc_destroyed)
+				mob_players[M] = music
+				music.players += src
+				if(fade_time > 0)
+					music.fade(1, fade_time)
+
+
+/datum/component/music_player/proc/stop_all(fade_time = 0)
+	for(var/_M in mob_players)
+		var/datum/music/sourced/music = mob_players[_M]
+		music.fade(0, fade_time)
+
+/datum/component/music_player/proc/remove_all()
+	for(var/_M in mob_players)
+		var/mob/M = _M
+		if(!M || !M.client)
+			continue
+		var/datum/music/sourced/music = mob_players[M]
+		music.players -= src
+	mob_players.len = 0
+
+/datum/component/music_player/Destroy()
+	. = ..()
+	STOP_PROCESSING(SSprocessing, src)
+	remove_all()
+	mob_players = null
+
+/datum/component/music_player/process()
+	if(enabled)
+		do_range_check()
+
+/datum/component/music_player/battle/process()
+	. = ..()
+	var/mob/M = parent
+	var/should_be_enabled = !M.stat && !M.client
+	if(enabled && !should_be_enabled)
+		remove_all(fade_in_time)
+	enabled = should_be_enabled
+
+/datum/music/sourced/battle
+	does_loop = TRUE
+
+/datum/music/sourced/battle/bubblegum
+	sound_file = 'sound/soundtrack/megafauna/bubblegum.ogg'
+	priority = 120
```

Comment:
There should be defines for these so the priority of each track is clear

Reply:
meh, thats just overdoing it

Reply:
No its not, we need to be able to easilly see the priority of other tracks in order to
a) Easilly change the other ones
b) Easilly insert the priority where we want
List things like this would be painful without defines (planes and layers)